feat: endpoint to update location for existing booking

Ryukemeister opened the pull request:
## What does this PR do?

Adds a v2 API endpoint to update the location of an existing booking. This supports CAL-6536 by letting orgs and teams edit conferencing details on booked events.

- Fixes CAL-6536

## Summary

- **New Endpoint**: `PATCH /v2/bookings/:bookingUid/location` to update booking location using typed location objects (address, link, phone, attendeeAddress, attendeePhone, attendeeDefined)
- **Requirements**: `cal-api-version: 2024-08-13` header, API key or access token, and `BOOKING_WRITE` permission
- **Guards**: Protected by ApiAuthGuard, PermissionsGuard, and BookingUidGuard
- **Authorization**: Only event type owners, hosts, team admins/owners, or org admins/owners can update booking locations (returns 403 Forbidden otherwise)
- **Response**: Returns 200 OK with the updated booking. If no location is sent, returns the current booking unchanged
- **Known Limitation**: Calendar event location is not updated yet; only the Cal.com booking is changed

## Updates since last revision

- Added authorization check in `BookingLocationService_2024_08_13` using `EventTypeAccessService.userIsEventTypeAdminOrOwner()` to ensure only authorized users can update booking locations
- Fixed E2E test to use valid location object format (object with `type` property instead of plain string)
- E2E tests use repository fixtures for creating test data to avoid authentication issues during setup

## Human Review Checklist

- [ ] Verify authorization logic in `apps/api/v2/src/ee/bookings/2024-08-13/services/booking-location.service.ts` correctly restricts access
- [ ] Confirm the 403 test case properly validates that unrelated users cannot update booking locations
- [ ] Note: Calendar events are intentionally NOT updated (documented limitation)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. N/A - API docs are auto-generated from decorators.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

1. Create a booking via the API or UI
2. Attempt to update the booking location using `PATCH /v2/bookings/:bookingUid/location` with a valid location object:
   ```json
   { "location": { "type": "link", "link": "https://example.com/meeting" } }
   ```
3. Verify the booking location is updated (200 OK)
4. Attempt to update with an unrelated user's token - should return 403 Forbidden
5. Attempt without authentication - should return 401 Unauthorized

---

Link to Devin run: https://app.devin.ai/sessions/d94d4bfdfcfa42a7b3033e40f433ee29
Requested by: rajiv@cal.com (@Ryukemeister)

Ryukemeister committed:
feat: init endpoint to update booking location

Ryukemeister committed:
chore: update controller

Ryukemeister committed:
chore: DTOs for update location endpoint

Ryukemeister committed:
chore: add update booking method

Ryukemeister committed:
fix: refactor

Ryukemeister committed:
feat: service update booking location

Ryukemeister committed:
chore: update platform libraries

Ryukemeister committed:
chore: update exports

Ryukemeister committed:
chore: e2e tests for updating booking location endpoint

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

Ryukemeister committed:
chore: update e2e tests for updating booking location endpoint

Ryukemeister committed:
chore: update bookings module

Ryukemeister committed:
fixup

Ryukemeister committed:
chore: update configs

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

Ryukemeister committed:
fixup: revert changes

Ryukemeister committed:
chore: remove unused handlers

Ryukemeister committed:
fix: update inputs

Ryukemeister committed:
refactor: booking location service

Ryukemeister committed:
fixup

Ryukemeister committed:
revert: unnecessary changes

Ryukemeister committed:
chore: cleanup

Ryukemeister committed:
fix: e2e tests

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

Ryukemeister committed:
chore: implement cubic feedback

supalarry commented on apps/api/v2/src/ee/bookings/2024-08-13/controllers/booking-location.controller.ts:
```diff
@@ -0,0 +1,58 @@
+  version: VERSION_2024_08_13_VALUE,
+})
+@UseGuards(PermissionsGuard)
+@DocsTags("Bookings / Location")
```
I think we can use `@DocsTags("Bookings")` so that this endpoint is together with other bookings endpoints.

supalarry commented on apps/api/v2/src/ee/bookings/2024-08-13/services/booking-location.service.ts:
```diff
@@ -0,0 +1,91 @@
+    return this.bookingsService.getBooking(existingBooking.uid, user);
+  }
+
+  getLocationValue(loc: BookingInputLocation_2024_08_13) {
```
Can we move this function under `updateLocation` and make it private since it's not gonna be used outside of this service for now? And we move it under it so code reads top to bottom aka updateBookingLocation -> updateLocation -> getLocationValue and we can also make `updateLocation` private I think.

supalarry commented on apps/api/v2/src/ee/bookings/2024-08-13/services/booking-location.service.ts:
```diff
@@ -0,0 +1,91 @@
+    const locationString = this.getLocationValue(location);
+    bookingLocation = locationString ?? bookingLocation;
+
+    const transformedLocation = this.inputService.transformLocation(location);
```
Maybe we can call this `const bookingFieldsLocation` ?

supalarry commented on packages/platform/types/bookings/2024-08-13/inputs/update-location.input.ts:
```diff
@@ -0,0 +1,135 @@
+  };
+
+  async validate(location: { type: string } | string) {
+    if (typeof location === "string") {
```
We can remove `if (typeof location === "string") {` because in `packages/platform/types/bookings/2024-08-13/inputs/location.input.ts` we have it because people could at first pass `booking.location` as a string and then we added specific objects aka `{type: phone, phone: x}` etc.

But since this is a new endpoint we only accept the objects from the start so its fine - we don't accept string location anymore. Does it make sense?

supalarry reviewed (dismissed):
Great stuff man! Have some review comments below let me know what you think 🙏

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

Ryukemeister committed:
chore: implement feedback

Ryukemeister committed:
fixup

supalarry commented on packages/platform/types/bookings/2024-08-13/inputs/location.input.ts:
```diff
@@ -157,14 +157,7 @@ class BookingInputLocationValidator_2024_08_13 implements ValidatorConstraintInt
-      return true;
-    }
-
+  async validate(location: { type: string }) {
```
I don't see this file changed at all in PR files

supalarry reviewed (commented):
Good stuff! I added 1 NIT small comment and then I have a question - if a location is updated then in the google calendar events the old one still persists, right? If so, we should add it to the update location endpoint docs to be clear about current state and then create linear ticket to update calendar event with the new location too.

Ryukemeister committed:
chore: more PR feedback

supalarry reviewed (dismissed):
Great work! 🚀

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

Ryukemeister committed:
fix: resolve merge conflicts

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

Ryukemeister committed:
fixup

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

ThyMinimalDev committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

devin-ai-integration[bot] committed:
fix: use repository fixture instead of API call in update-booking-location e2e test

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

devin-ai-integration[bot] committed:
fix: use repository fixture for createBooking helper in update-booking-location e2e test

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
fix: use valid location object format in 403 test for update-booking-location

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

devin-ai-integration[bot] committed:
fix: add authorization check to booking location update endpoint

The updateBookingLocation endpoint was missing an authorization check,
allowing any authenticated user to update any booking's location.

This commit adds the authorization check using EventTypeAccessService
to verify that the user is the event type owner, host, team admin/owner,
or org admin/owner before allowing the location update.

This fixes the failing E2E test that expected 403 Forbidden when an
unrelated user attempts to update a booking location.

Co-Authored-By: rajiv@cal.com <sahalrajiv6900@gmail.com>

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for

Ryukemeister committed:
Merge branch 'main' into rajiv/cal-6536-for-orgsteams-add-a-way-to-edit-conferencing-app-for
